### PR TITLE
Roadmap

### PR DESCRIPTION
**Pull request by kauffj**
If you don't know where you're going then any road will take you there

### CODE REVIEW

File: controller/action/ContentActions.class.php [hunk 1]
```diff
@@ -192,6 +191,33 @@ public static function executeBounty(string $slug = null): array
     ]];
   }
 
+  public static function executeRoadmap()
+  {
+    $cache = !isset($_GET['nocache']);
```
Comment (lyoshenka) [inline controller/action/ContentActions.class.php]: Do not do this. Use `Request::getParam()` instead

File: lib/cache/Apc.class.php [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+/**
+ * Created by PhpStorm.
```
Comment (lyoshenka) [inline lib/cache/Apc.class.php]: please remove this comment. you should probably get phpstorm to stop inserting them

File: lib/thirdparty/Asana.class.php [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+class Asana
+{
+  protected static $curlOptions = ['json_response' => true, 'cache' => true];
+
+  public static function listRoadmapTasks($cache = true)
+  {
+    /*
+     *   return static::get('/projects');
+            [55] => Array
```
Comment (lyoshenka) [inline lib/thirdparty/Asana.class.php]: please delete this

File: lib/thirdparty/Github.class.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
+<?php
+
+class Github
```
Comment (lyoshenka) [inline lib/thirdparty/Github.class.php]: i support this refactor :+1:

File: lib/thirdparty/Prefinery.class.php [hunk 5]
```diff
@@ -17,10 +17,10 @@ class Prefinery
...
-    'json_data' => true
+    'json_data' => true,
+    'json_response' => true
   ];
-
-
+  
   public static function findUser($emailOrId, $useApc = true)
   {
     $apcEnabled = extension_loaded('apc') && ini_get('apc.enabled');
```
Comment (lyoshenka) [inline lib/thirdparty/Prefinery.class.php]: didnt you add a function for this?

File: lib/tools/Curl.class.php [hunk 6]
```diff
@@ -45,13 +48,28 @@ public static function doCurl($method, $url, $params = [], $options = [])
...
     }
 
+    $options = array_merge($defaults, $options);
+
+    if (!Apc::isEnabled())
+    {
+      $options['cache'] = false;
+    }
+
+    if ($options['cache'])
```
Comment (lyoshenka) [inline lib/tools/Curl.class.php]: I don't think this cache stuff belongs in the Curl class. happy to discuss if you disagree.

**Comment by lyoshenka** [comment]
You're gonna have to rebase this onto master too, btw.

File: lib/tools/Curl.class.php (hunk 6 above)
Comment (kauffj) [inline lib/tools/Curl.class.php]: If the choices are either:
1) Have an APC wrapper version of cURL
2) Allow cURL to cache

I pick the second. Feel free to refactor or suggest something else.

**Comment by kauffj** [comment]
@lyoshenka I made all requested changes but cURL refactor. Let's get this up.

File: lib/tools/Curl.class.php [hunk 7]
```diff
@@ -27,16 +28,18 @@ public static function put($url, $params = [], $options = [])
   public static function doCurl($method, $url, $params = [], $options = [])
   {
     $defaults = [
+      'cache'            => false,
       'headers'          => [],
       'verify'           => true,
-      'timeout'          => 5,
+      'timeout'          => 10, //5 was timing out on /roadmap for Asana API
```
Comment (lyoshenka) [inline lib/tools/Curl.class.php]: the default should still be 5 (or even lower). you can bump the option up to 10 for asana requests if you want

**Pull request merged**